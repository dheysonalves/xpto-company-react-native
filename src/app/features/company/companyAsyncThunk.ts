import { createAsyncThunk } from "@reduxjs/toolkit";
import { CompanyService } from "../../../services";

const fetchAllCompanies = createAsyncThunk(
	"company/fetchAllCompanies",
	async () => {
		const response = await CompanyService.getAllCompanies();
		return response;
	}
);

export {
	fetchAllCompanies
}
