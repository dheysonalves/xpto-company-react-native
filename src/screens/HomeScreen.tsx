import React from 'react';
import { Platform, SafeAreaView, StyleSheet, View, Text } from "react-native";
import { StatusBar } from "expo-status-bar";
import CardList from "../components/presentation/CardList/CardList.component";

import { TextInput, Button } from "../components/primitives";
import { useNavigation } from '@react-navigation/native';

export default function HomeScreen() {
	const [search, updateSearch] = React.useState('');
	const { navigate } = useNavigation();

	return (
		<SafeAreaView style={styles.container}>
			<StatusBar style={Platform.OS === "ios" ? "light" : "auto"} />
			<Text style={styles.title}>BANQI PROJECT</Text>
			<TextInput
				title="Search"
				dataValue={search}
				onHandleDataValue={updateSearch}
				hasError={false}
				keyboardType="web-search"
			/>
			<CardList />
			<View style={styles.buttonWrapper}>
				<Button
					text="ADD COMPANY"
					handleSubmit={() => navigate("NewCompanyView")}
				/>
			</View>
		</SafeAreaView>
	);
}

const styles = StyleSheet.create({
	container: {
		flex: 1,
		justifyContent: "center",
		backgroundColor: "#fff",
		padding: 24,
	},
	title: {
		fontSize: 42,
		fontWeight: "bold",
		color: "#0a0a0a",
		textAlign: "center",
		textTransform: "uppercase",
	},
	buttonWrapper: {
		shadowColor: "#000",
		shadowOffset: {
			width: 0,
			height: 2,
		},
		shadowOpacity: 0.25,
		shadowRadius: 3.84,

		elevation: 5,
		marginTop: 24,
	},
});
