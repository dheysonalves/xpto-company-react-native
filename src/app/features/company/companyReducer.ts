import { createSlice, isFulfilled, isRejected, isPending, PayloadAction } from "@reduxjs/toolkit";
import { ErrorType } from "../../../@types/requests";
import { ICompanyInformation } from "../../../services/companyService";
import { fetchAllCompanies } from './companyAsyncThunk';

export interface CompanyState {
	companies: ICompanyInformation[];
	company: ICompanyInformation | null;
	loading: 'loading' | 'idle';
	error: ErrorType;
	isError: boolean;
}

const initialState: CompanyState = {
	companies: [],
	company: null,
	loading: 'idle',
	error: undefined,
	isError: false,
};

const isAFulfilledAction = isFulfilled(fetchAllCompanies);
const isAPendingAction = isPending(fetchAllCompanies);
const isARejectedAction = isRejected(fetchAllCompanies);

export const companySlice = createSlice({
	name: "company",
	initialState,
	reducers: {
		setCompanyData: (
			state: CompanyState,
			{ payload }: PayloadAction<ICompanyInformation | null>
		) => {
			state.company = payload;
		},
		setLoadingState: (
			state: CompanyState,
			{ payload }: PayloadAction<"loading" | "idle">
		) => {
			state.loading = payload;
		},
	},
	extraReducers(builder) {
		builder
			.addCase(fetchAllCompanies.fulfilled, (state, { payload }) => {
				state.companies = payload;
			})
			.addMatcher(isAFulfilledAction, (state) => {
				state.loading = "idle";
			})
			.addMatcher(isAPendingAction, (state) => {
				state.loading = "loading";
				state.error = undefined;
				state.isError = false;
			})
			.addMatcher(isARejectedAction, (state, { error }) => {
				state.loading = "idle";
				state.error = error;
				state.isError = true;
			});
	},
});

export const { setLoadingState } = companySlice.actions;

export default companySlice.reducer;
