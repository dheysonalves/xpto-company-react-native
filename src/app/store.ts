import { configureStore } from '@reduxjs/toolkit'
import companyReducer from './features/company/companyReducer'

export const store = configureStore({
	reducer: {
		company: companyReducer,
	},
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
