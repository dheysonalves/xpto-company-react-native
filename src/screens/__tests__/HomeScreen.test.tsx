import React from "react";

import HomeScreen from "../HomeScreen";
import renderer from "react-test-renderer";
import { cleanup } from "@testing-library/react-native";
import { Provider } from "react-redux";
import { RootState, store } from "../../app/store";
import { NavigationContainer } from "@react-navigation/native";
import { fetchAllCompanies } from "../../app/features/company/companyAsyncThunk";
import Companies from "../../mocks/companies";

afterEach(cleanup);
jest.useFakeTimers();

const renderWithContext = (element: JSX.Element) => {
	return (
		<NavigationContainer>
			<Provider store={store}>{element}</Provider>
		</NavigationContainer>
	);
};

describe("Home Screen test", () => {
	it("should render the home screen", () => {
		const tree = renderer.create(renderWithContext(<HomeScreen />)).toJSON();

		expect(tree).toMatchSnapshot();
	});
	it("it should get all the companies", async () => {
		const dispatch = jest.fn();
		const state: RootState = {
			company: {
				companies: [],
				company: {
					address: {
						city: "",
						complement: "",
						neighborhood: "",
						number: "",
						state: "",
						street: "",
						zip: "",
					},
					cnpj: "",
					description: "",
					name: "",
					createdAt: "",
					id: "",
					logo: "",
				},
				error: {},
				isError: false,
				loading: "idle",
			},
		};
		const thunk = fetchAllCompanies();
		await thunk(dispatch, () => state, undefined);
		const { calls } = dispatch.mock;
		expect(calls).toHaveLength(2);
		expect(calls[0][0].type).toEqual("company/fetchAllCompanies/pending");
		expect(calls[1][0].type).toEqual("company/fetchAllCompanies/fulfilled");
		expect(calls[1][0].payload).toEqual(Companies);
	});
});
