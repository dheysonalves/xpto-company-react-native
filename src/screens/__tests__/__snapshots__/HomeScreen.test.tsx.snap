// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen test should render the home screen 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "flex-start",
      "padding": 24,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "color": "#0a0a0a",
          "fontSize": 42,
          "fontWeight": "bold",
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
    >
      BANQI PROJECT
    </Text>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "margin": 12,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 18,
                "fontWeight": "700",
                "letterSpacing": 0.6,
                "marginVertical": 4,
              },
              Object {
                "color": "hsl(184, 14%, 56%)",
              },
            ]
          }
        >
          Search
        </Text>
      </View>
      <View
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <TextInput
          keyboardType="web-search"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          selectionColor="hsl(183, 100%, 15%)"
          style={
            Array [
              Object {
                "backgroundColor": "hsl(189, 41%, 97%)",
                "borderRadius": 5,
                "color": "hsl(183, 100%, 15%)",
                "fontSize": 24,
                "fontWeight": "700",
                "letterSpacing": 0.6,
                "paddingHorizontal": 16,
                "paddingVertical": 8,
              },
              null,
              null,
            ]
          }
          testID="textInputSearch"
          value=""
        />
      </View>
    </View>
  </View>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    testID="cardList"
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
